Write matching games appendix (#85)
daffidwilde/thesis: appendix/matching/main.tex

\chapter{An introduction to matching games}
\label{app:matching}

\graphicspath{{appendix/matching/paper/img/}}

Matching games form a part of game theory that were formally introduced by Gale
and Shapley in their seminal work~\cite{Gale1962}. These games allow for the
allocation of resources and partnerships in a mathematically fair way.

> Typically, a matching game is defined by two sets of players (called parties)

Reviewer: ```suggestion
Typically, a matching game is defined by two sets of players (referred to as parties)
```

that each have preferences over at least some of the elements of the other set.
The objective of the game is then to find a mapping between the sets of players
in which everyone is \emph{happy enough} with their match(es).

This appendix does not contain any novel mathematics, but it does offer an
introduction to matching games and their variants. Studying this branch of
mathematics has contributed to a significant amount of the research conducted
for this thesis, hence its inclusion here. That research has culminated in the
development of a Python library for solving various matching games, \matching.
Among other uses, the \matching\ library proves instrumental in the practical
implementation of the novel method described in Chapter~\ref{chp:kmodes}.

The \matching\ library has been developed as a research tool and adheres to the
best practices discussed in Chapter~\ref{chp:intro}. The current version of

> Matching has also been archived on Zenodo under~\doi{10.5281/zenodo.3931026}.

Reviewer: Do you want to say `current` or should it be more pointing at the version used in this thesis? (Or did you use two different versions -- biosci versus initialisation?)

Author: The HR and SA implementations are the same in each case, but I think they use slightly different versions of `matching`. I believe that link goes to the `matching` Zenodo page (which redirects to the most recent version)

I will point at the correct versions in the k-modes chapter and biosci appendix, though

Reviewer: :+1:

Author: ✅ 1e6978c

Along with the source code being modularised and fully tested (using example,
integration and property-based unit tests) with 100\% coverage, the library is
documented extensively. Like the \edo\ library developed for the work in
Chapter~\ref{chp:edo}, the \matching\ documentation is hosted online at

> \href{https://matching.readthedocs.io}{\nolinkurl{matching.readthedocs.io}}.

Reviewer: What does `\nolinkurl` do?

Author: `\nolinkurl` just renders the text within as if it were a URL but without a destination. I'm not sure why but `hyperref` won't pick up on some changes I've made to how URLs should be formatted so this is my workaround.

A cleaner solution would be to make my own `url` command:

```tex
\renewcommand{\myurl}[1]{%
    \href{https://#1}{\nolinkurl{#1}
}
```

... but I don't think I do this anywhere else so 🤷‍♂️

Reviewer: Neat. TIL.

The documentation has been written to maximise its effect as a resource for
learning about matching games as well as for the software itself. Furthermore,

> the library is registered on the Python Package Index and is \pip-installable.

Reviewer: ```suggestion
the library is registered on the Python Package Index and is installable using standard python practices.
```

What do you think?

Author: This is much better. I couldn't put my finger on how to say it otherwise 😅

\begin{listing}
\begin{usagesh}
> pip install matching
\end{usagesh}
\caption{Installing the \matching\ library via \pip}
\end{listing}

Matching games have applications in many fields where relationships between

> rational agents must be managed. Some example applications include: being able

Reviewer: ```suggestion
rational agents must be arranged. Some example applications include: being able
```

Not sure.

to inform on healthcare finance policy~\cite{Agarwal2017}; helping to reduce the
complexity of automated wireless communication networks~\cite{Bayat2016}; and
education infrastructure~\cite{Chiarandini2019}. Thus, having access to software
implementations of algorithms that are able to solve such games is essential.

> The only current adversary to \matching\ is \matchingr~\cite{Tilly2018}.

Reviewer: ```suggestion
The only current software alternative to \matching\ is \matchingr~\cite{Tilly2018}.
```

\matchingr\ is a package written in C++ with an R interface and its content
overlaps well with that of \matching. However, the lack of a Python interface

> makes it less relevant to researchers and other users as Python's popularity

Reviewer: Can we think of a way to reword this? I know what you're saying but a lot of R people would say that a Python library is irrelevant...

Author: There definitely is a better way of writing this; I lifted this paragraph directly from the JOSE paper on the matching library and I wasn't the biggest fan then

grows both in academia and industry.

At the time of writing, the \matching\ library offers facilities to handle and

> solve four types of matching game:

Reviewer: ```suggestion
solve four types of matching games:
```

\begin{itemize}
    \item The stable marriage problem (SM)~\cite{Gale1962};
    \item the hospital-resident assignment problem
        (HR)~\cite{Gale1962,Roth1984};
    \item the student-project allocation problem (SA)~\cite{Abraham2007}; and
    \item the stable roommates problem (SR)~\cite{Irving1985}.
\end{itemize}

This appendix goes through the details of the games for SM and HR, the latter
of which is used in Chapter~\ref{chp:kmodes}. A further piece of work conducted
during the production of this thesis that uses SA is provided in

> Appendix~\ref{app:biosci}.

Reviewer: ```suggestion
Appendix~\ref{app:biosci}.  % TODO Remember to add reference (if we publish it)
```

    R)\). A pair \((s, r) \in S \times R\) is said to \emph{block} \(M\) if:

    \begin{itemize}
        \item \(s\) and \(r\) are not matched by \(M\), i.e. \(M(s) \neq r\);
        \item \(s\) prefers \(r\) to \(M(s) = r'\); and
        \item \(r\) prefers \(s\) to \(M^{-1}(r) = s'\).
    \end{itemize}

    A matching \(M\) is said to be \emph{stable} if it has no blocking pairs,
    and \emph{unstable} otherwise.
\end{definition}

This final definition envelopes the critical differences between the various
matching games in existence. Despite their differences, however, the spirit is
the same: a pair of players blocks a matching if their envy is \emph{mutually
rational}. Irrational envy would be where one player wishes to be matched to
another over their current match but the other player does not (or cannot)
reciprocate. The social outcome of acting irrationally in SM is that a player
would be betraying their partner for another player, thus destabilising the
group, without any reward of a `better' partner.

Consider the game of size three shown in Figure~\ref{fig:sm_matching} as an
edgeless graph with suitors on the left and reviewers on the right. This

> representation of a matching game finds it origin in the bipartite matching

Reviewer: ```suggestion
representation of a matching game finds its origin in the bipartite matching
```

problems of graph theory. Beside each vertex is the name of the player and their
associated ranking of the complementary set’s elements.

\begin{figure}[htbp]
    \centering
    \input{appendix/matching/paper/docs/tex/sm_matching}
    \caption{A game of size three}\label{fig:sm_matching}
\end{figure}

In this representation, a matching \(M\) creates a bipartite graph where an edge
between two vertices (players) indicates that they are matched by \(M\).
Figure~\ref{fig:sm_unstable} shows an example of a valid matching.

\begin{figure}[htbp]
    \centering
    \input{appendix/matching/paper/docs/tex/sm_unstable}
    \caption{An unstable matching to the game}\label{fig:sm_unstable}
\end{figure}

In this matching, players \(A\), \(C\) and \(F\) are all matched to their
favourite player while \(B\), \(D\) and \(E\) are matched to their least
favourite. In particular, \(B\) and \(D\) form a blocking pair since they would
both rather be matched to one another than their current match. Hence, this

> matching is unstable. As an attempt to rectify this instability, let the matches
> for the first two rows swap, as shown in Figure~\ref{fig:sm_stable}. This move

Reviewer: ```suggestion
matching is unstable. As an attempt to rectify this instability, swap the matches
for the first two rows, as shown in Figure~\ref{fig:sm_stable}. This move
```

does not form another blocking pair despite \(A\) having a worse match since
\(D\) ranks \(A\) at the bottom of its preference list. Therefore, the envy
exhibited by \(A\) is not reciprocated, and the matching is stable.

\begin{figure}[htbp]
    \centering
    \input{appendix/matching/paper/docs/tex/sm_stable}
    \caption{A stable, suitor-optimal solution to the game}\label{fig:sm_stable}
\end{figure}

Upon closer inspection of this matching, it appears that no suitor can improve
on their current match without forming a blocking pair. In fact, the only suitor
improvement would be for \(A\) and \(D\) to be matched again. This kind of
stable matching is called \emph{suitor-optimal}. Similarly, no reviewer can
improve their match without forming a blocking pair and so this matching is also
\emph{reviewer-optimal}.

Finding a party-optimal, stable matching to an instance of a matching game is
referred to as \emph{solving} the game. When there are only a handful of players
to deal with, solving a game (or even finding a party-suboptimal, stable
matching) is relatively straightforward with pen, paper and some time. However,

> solving the example above in two steps was little more than a coincidence,
> really. In the seminal paper on matching games~\cite{Gale1962}, Gale and Shapley

Reviewer: ```suggestion
solving the example above in two steps was little more than a coincidence. 
In the seminal paper on matching games~\cite{Gale1962}, Gale and Shapley
```

presented an algorithm for finding a unique, stable and suitor-optimal matching
to any instance of SM. This algorithm has since become known as the Gale-Shapley
algorithm, and is given in Algorithm~\ref{alg:stable_marriage}. The matching

> this algorithm produces is that shown in Figure~\ref{fig:sm_stable}.

Reviewer: ```suggestion
this algorithm produces is shown in Figure~\ref{fig:sm_stable}.
```

ranking them all the same: as unacceptable. Further to these extensions, HR has
given rise to its own contextual problems. One of these is allowing for
couples in the resident party. Studies on this problem have shown that no stable
matching is guaranteed to exist, and so a related, NP-hard problem is considered
instead where the objective is to identify an almost-stable matching that
minimises the number of blocking pairs~\cite{Manlove2016}.

Another method used to construct preference lists is to discount the preference
lists presented by players. For instance, where acceptability of another player
is the only criterion, binary preferences (i.e.\ incomplete preference lists
with ties) can create games that are invulnerable to manipulative players'
strategies~\cite{Bogomolnaia2004}. This approach can be adapted to cater for
larger games, such as student-school allocation (a special case of HR). In this
scenario, each student submits a set of acceptable schools and the schools form
strict rankings of the students. The result of this is a simpler game (in the
practical sense) and a reduction in the set of possible stable
matchings~\cite{Haeringer2014,Haeringer2019}.

> 
> 

Reviewer: ```suggestion
```
